### PR TITLE
Added "Sort By" function under Window Menu

### PR DESCRIPTION
**Pull request by deebash**
Added sorting function under Window Menu for issue #10393

### CODE REVIEW

**Comment by Yaron10** [comment]
@deebash,

Thank you for this PR. 
I've tested it and it works like a charm. Great work. 👍 

Allow me some comments:
1. Wouldn't it be nicer if the "Sort By" menu appeared *after* the list of open documents?
List of open documents.
Separator.
Sort By.
Windows...

2. Isn't it possible to adapt and use the existing Sort functions?

3. If this PR is accepted, you'd be asked to provide localization for the new items.

**Comment by deebash** [comment]
Thanks @Yaron10 

1. Sure, I'll change that. I placed it on top so that if large number of documents are opened users needs to scroll to bottom all the way down to reach this menu item. If you prefer the way that you mentioned, I'll do the change.

2. Actually I used the existing sort function code, but removed few lines which is related to list view processing in Windows.... Dialog box.

3. Sure, I'll do that.

**Comment by Yaron10** [comment]
@deebash,

> I placed it on top so that if large number of documents are opened...

The number of open documents is limited to 10.
If @donho approves, I don't see why it shouldn't be changed to 20.

Thanks again.

File: PowerEditor/src/NppCommands.cpp
Comment (mere-human) [inline PowerEditor/src/NppCommands.cpp:3560]: Underscore at the beginning is usually used only for member variables.
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md#member-variables

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp
Comment (mere-human) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:953]: You can also use uniform initialization to avoid uninitialized members. Like this: `nmdlg = {};`

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp
Comment (mere-human) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:959]: In order to avoid a potential memory leak, you can use a RAII approach. Example:
```
std::vector<UINT> items(count);
nmdlg.Items = items.data();
```

**Comment by deebash** [comment]
I did the changes as mentioned by @mere-human. @donho as per contribution guidelines, its better to give very less commit in a PR for easy code review. So, If you accept this PR, I'll add localisation text and change the menu item position based upon your suggestion and I'll add single commit with both changes.

**Comment by Yaron10** [comment]
@deebash,

> List of open documents.
Separator.
Sort By.
Windows...

A few months ago I modified some code in `WindowsDlg.cpp` for my personal use.
Amongst other things, I've added a separator between the list of open documents and the Windows... command.

I did that by adding the following lines in `WindowsDlg.cpp` after the `for ( ; id <= IDM_WINDOW_MRU_LIMIT; ++id)` loop.
```
::RemoveMenu(hMenu, 0, MF_SEPARATOR);
::InsertMenu(hMenu, IDM_WINDOW_WINDOWS, MF_SEPARATOR | MF_BYCOMMAND, 0, NULL);
```
Could you please let me know if you have a better solution?

Thank you.

**Comment by deebash** [comment]
> Could you please let me know if you have a better solution?
@Yaron10 If you want to add menu item at the bottom, then you can use -1 as position. It'll append to bottom of the menu

**Comment by Yaron10** [comment]
@deebash,

I'll wait for your final PR. Thank you.

**Comment by deebash** [comment]
Hi @donho, waiting for your approval on the menu position.

**Comment by deebash** [comment]
@donho  @mere-human changes has been done and commit added.

**Comment by mere-human** [comment]
Looks good. I haven't tested it locally yet.

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3626]: Indent problem?

Comment (Mateos81) [reply PowerEditor/src/NppCommands.cpp:3626]: The usual mixed tabs/spaces

Comment (deebash) [reply PowerEditor/src/NppCommands.cpp:3626]: I"m using Visual studio code, in there indent is fine. After pushing, this indent error happened.

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:986]: Refactoring: I would add a new method here with the following code in `WindowsDlg.cpp`:
```
void WindowsDlg::sort(int columnID, bool reverseSort)
{
	refreshMap();
	_currentColumn = columnID;
	_reverseSort = reverseSort;
	stable_sort(_idxMap.begin(), _idxMap.end(), BufferEquivalent(_pTab, _currentColumn, _reverseSort));
}
```

Then use the sort() method on the following methods, in `WindowsDlg.h`:
```
void WindowsDlg::sortFileNameASC() {
	sort(0, false);
}

void WindowsDlg::sortFileNameDSC() {
	sort(0, true);
}

void WindowsDlg::sortFilePathASC() {
	sort(1, false);
}

...
```

Note that the coding conversion (braces position) should be respected:
in cpp:
```
void aClass::method1()
{
	//  codes here
}
```
in h:
```
class aClass
{
public:
	void method1();
	int method1() {
		return _x; // only one line code can be placed in *.h
	}

private:
	int _x;
}
```

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.h
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.h:60]: Put the one line definition code here, with the new method `sort()`.

**Review by donho — CHANGES_REQUESTED** [review]
1. code refactoring.
2. solve the indent problem (in comment).

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3617]: Indent problem

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3608]: Indent problem

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3599]: Indent problem

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3590]: Indent problem

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3572]: Indent problem

File: PowerEditor/src/NppCommands.cpp
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:3581]: Indent problem

**Comment by donho** [comment]
@Yaron10 
> The number of open documents is limited to 10.
If @donho approves, I don't see why it shouldn't be changed to 20.

I will consider it - could you create an issue for that please?

**Comment by donho** [comment]
@deebash 
There are 2 missing stuffs:
1. Localization (that you have mentioned).
2. The possibility to add/change the shortcut of these command: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Parameters.cpp#L65

You can do them in this PR or in the next PR.
If you do them in the next PR, please create a new branch for each PR:
`Create a new branch for each PR. Make sure your branch name wasn't used before - you can add date (for example patch3_20200528) to ensure its uniqueness.`
Ref: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md#guidelines-for-pull-requests

Thank you

**Comment by deebash** [comment]
@donho I'll do in this PR itself, I'll give commit in few hours. And thank you for the brief comments.

File: PowerEditor/installer/nativeLang/english.xml
Comment (donho) [inline PowerEditor/installer/nativeLang/english.xml:655]: Please add "Sort By" before all the entries. For example:
```xml
                    <Item id="11002" name="Sort By Name A to Z"/>
                    <Item id="11003" name="Sort By Name Z to A"/>
                    <Item id="11004" name="Sort By Path A to Z"/>
                    <Item id="11005" name="Sort By Path Z to A"/>
                    <Item id="11006" name="Sort By Type A to Z"/>
...
```
because these  entries name will appear in the shortcut mapper without "Sort By" menu. With "Sort By" in front of each command, users won't be confused.

**Review by donho — CHANGES_REQUESTED** [review]
Add "Sort By" before all the entries of MainCommandNames in ShortcutMapper.

File: PowerEditor/installer/nativeLang/english.xml
Comment (deebash) [reply PowerEditor/installer/nativeLang/english.xml:655]: @donho Modifying english.xml not taking effect when notepad++ launches, it works only after changes back and forth the language in Preferences. updateCommandShortcuts() is triggering but it's not updating the strings. Is there any step that I'm missing?

**Comment by deebash** [comment]
@donho I added Sort By prefix before each entries. But changes are taking effect only after language change in preferences. Is this the actual behaviour? It's happening for all entries.

**Comment by donho** [comment]
@deebash 
> I added Sort By prefix before each entries. But changes are taking effect only after language change in preferences. Is this the actual behaviour?

What do you mean by "changes are taking effect only after language change in preferences." ?

**Review by donho — APPROVED** [review]

File: PowerEditor/src/Parameters.cpp
Comment (donho) [inline PowerEditor/src/Parameters.cpp:416]: > changes are taking effect only after language change in preferences.

I guess I do understand what you said now - my bad, I forgot these entries.
I'll do a PR to fix them (and you will understand why this behaviour you describe above).

File: PowerEditor/src/localization.cpp
Comment (donho) [inline PowerEditor/src/localization.cpp:373]: Normally, it shouldn't be here this block.

**Comment by Yaron10** [comment]
@deebash & @donho,

👍 
Great work. Thank you.

File: PowerEditor/src/Parameters.cpp
Comment (deebash) [reply PowerEditor/src/Parameters.cpp:416]: Thank you.

**Comment by Yaron10** [comment]
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11271.
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11272.

**Pull request closed without merging**
